prefect_kubernetes/jobs.py

```diff
@@ -299,3 +309,197 @@ def kubernetes_orchestrator():
             namespace=namespace,
             **kube_kwargs,
         )
+
+
+class KubernetesJobRun(JobRun[Dict[str, Any]]):
+    """A run of a Kubernetes job.
+
+    Attributes:
+        _kubernetes_job: The Kubernetes job this run is for.
+    """
+
+    def __init__(self, kubernetes_job: "KubernetesJob"):
+        self._kubernetes_job = kubernetes_job
+        self.pod_logs = {}
+
+    async def wait_for_completion(self):
+        """Waits for the job to complete.
+
+        Raises:
+            RuntimeError: If the Kubernetes job fails.
+        """
+        with self._kubernetes_job.credentials.get_client(
+            "batch"
+        ) as batch_v1_client, self._kubernetes_job.credentials.get_client(
+            "core"
+        ) as core_v1_client:
+            completed = False
+
+            while not completed:
+                latest_v1_job = await run_sync_in_worker_thread(
+                    batch_v1_client.read_namespaced_job_status,
+                    name=self._kubernetes_job.v1_job.metadata.name,
+                    namespace=self._kubernetes_job.namespace,
+                    **self._kubernetes_job.api_kwargs,
+                )
+                print(latest_v1_job.status)
+
+                v1_pod_list = await run_sync_in_worker_thread(
+                    core_v1_client.list_namespaced_pod,
+                    namespace=self._kubernetes_job.namespace,
+                    label_selector=(
+                        f"controller-uid="
+                        f"{latest_v1_job.metadata.labels['controller-uid']}"
+                    ),
+                    **self._kubernetes_job.api_kwargs,
+                )
+                for pod in v1_pod_list.items:
+                    pod_name = pod.metadata.name
+
+                    if pod.status.phase == "Pending" or pod_name in self.pod_logs:
+                        continue
+
+                    self.logger.info(f"Capturing logs for pod {pod_name}.")
+
+                    self.pod_logs[pod_name] = await run_sync_in_worker_thread(
+                        core_v1_client.read_namespaced_pod_log,
+                        namespace=self._kubernetes_job.namespace,
+                        name=pod_name,
+                        container=latest_v1_job.spec.template.spec.containers[0].name,
+                        **self._kubernetes_job.api_kwargs,
+                    )
+
+                if latest_v1_job.status.active:
+                    await sleep(self._kubernetes_job.interval_seconds)
+                elif latest_v1_job.status.failed:
+                    raise RuntimeError(
+                        f"Job {latest_v1_job.metadata.name} failed, check the "
```

Comment:
I wonder if this should be in fetch_result?

Reply:
Maybe we can implement something like https://github.com/PrefectHQ/prefect-dbt/pull/101/files#diff-826f2d10c2bbb1fb21d3f9003e437a53594ab7b22f01e4a75b8a14ab7c9b4fedR687-R730 here to log the status changes?

Also, I think we need to implement timeout_seconds.